dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/CompilationVisitor.java

```diff
@@ -0,0 +1,860 @@
+        
+        Register resultRegister = e.getResultRegister();
+        IExpr address = e.getAddress();
+
+        Register dummyReg = new Register(null, resultRegister.getThreadId(), resultRegister.getPrecision());
+		Load load = newRMWLoad(dummyReg, address, null);
+        load.addFilters(Tag.TSO.ATOM);
+
+        RMWStore store = newRMWStore(load, address, resultRegister, null);
+        store.addFilters(Tag.TSO.ATOM);
+
+        Local updateReg = newLocal(resultRegister, dummyReg);
+
+        return eventSequence(
+                load,
+                store,
+                updateReg
+        );
+	}
+
+	@Override
+	public List<Event> visit(AtomicAbstract e) {
+		throw new UnsupportedOperationException("Compilation to " + target + 
+				" is not supported for " + getClass().getName());
+	}
+
+	@Override
+	public List<Event> visit(AtomicCmpXchg e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr address = e.getAddress();
+		ExprInterface value = e.getMemValue();
+		String mo = e.getMo();
+		IExpr expectedAddr = e.getExpectedAddr();
+        int threadId = resultRegister.getThreadId();
+		int precision = resultRegister.getPrecision();
+
+		List<Event> events;
+
+        // These events are common to all compilation schemes.
+        // The difference of each architecture lies in the used Store/Load to/from <address>
+        // and the fences that get inserted
+		Register regExpected = new Register(null, threadId, precision);
+        Register regValue = new Register(null, threadId, precision);
+        Load loadExpected = newLoad(regExpected, expectedAddr, null);
+        Store storeExpected = newStore(expectedAddr, regValue, null);
+        Label casFail = newLabel("CAS_fail");
+        Label casEnd = newLabel("CAS_end");
+        Local casCmpResult = newLocal(resultRegister, new Atom(regValue, EQ, regExpected));
+        CondJump branchOnCasCmpResult = newJump(new Atom(resultRegister, NEQ, IConst.ONE), casFail);
+        CondJump gotoCasEnd = newGoto(casEnd);
+
+        switch(target) {
+            case NONE:
+            case TSO: {
+                Load loadValue = newRMWLoad(regValue, address, mo);
+                Store storeValue = newRMWStore(loadValue, address, value, mo);
+
+                events = eventSequence(
+                        // Indentation shows the branching structure
+                        loadExpected,
+                        loadValue,
+                        casCmpResult,
+                        branchOnCasCmpResult,
+                            storeValue,
+                            gotoCasEnd,
+                        casFail,
+                            storeExpected,
+                        casEnd
+                );
+                break;
+            }
+            case POWER:
+            case ARM8: {
+                String loadMo = Tag.ARMv8.extractLoadMoFromCMo(mo);
+                String storeMo = Tag.ARMv8.extractStoreMoFromCMo(mo);
```

Comment:
It is kind of ugly that we use ARMv8 tags like this even for Power. But I think this is an orthogonal issue to this PR.

Reply:
Using such tags should not have any effect with `power.cat`. I think this ugly code is a co-product of having a single match statement in the switch for `ARM8` and `POWER`. Actually, we ended up converting things to power fences based on the extracted `loadMo` and `storeMo`, so we can achieve the same reasoning directly about the `Tag.C11`.

But as you said, this is orthogonal to this PR and I think we can treat in a different one where we implement one visitor per architecture.